src/js/tools/sentryMetroSerializer.ts

```diff
@@ -0,0 +1,167 @@
+    // eslint-disable-next-line no-console
+    console.log('info ' + `Bundle Debug ID: ${debugId}`);
+
+    const debugIdComment = `${DEBUG_ID_COMMENT}${debugId}`;
+    const indexOfSourceMapComment = bundleCode.lastIndexOf(SOURCE_MAP_COMMENT);
+    const bundleCodeWithDebugId =
+      indexOfSourceMapComment === -1
+        ? // If source map comment is missing lets just add the debug id comment
+          `${bundleCode}\n${debugIdComment}`
+        : // If source map comment is present lets add the debug id comment before it
+          `${bundleCode.substring(0, indexOfSourceMapComment) + debugIdComment}\n${bundleCode.substring(
+            indexOfSourceMapComment,
+          )}`;
+
+    const bundleMap: SourceMap = JSON.parse(bundleMapString);
+    // For now we write both fields until we know what will become the standard - if ever.
+    bundleMap['debug_id'] = debugId;
+    bundleMap['debugId'] = debugId;
+
+    return {
+      code: bundleCodeWithDebugId,
+      map: JSON.stringify(bundleMap),
+    };
+  };
+};
+
+/**
+ * This function is expected to be called after serializer creates the final bundle object
+ * and before the source maps are generated.
+ *
+ * It injects a debug ID into the bundle and returns the modified bundle.
+ *
+ * Access it via `options.sentryBundleCallback` in your custom serializer.
+ */
+function createSentryBundleCallback(debugIdModule: Module<VirtualJSOutput> & { setSource: (code: string) => void }) {
+  return (bundle: Bundle) => {
+    const debugId = calculateDebugId(bundle);
+    debugIdModule.setSource(injectDebugId(debugIdModule.getSource().toString(), debugId));
+    bundle.pre = injectDebugId(bundle.pre, debugId);
+    return bundle;
+  };
+};
+
+function addDebugIdModule(preModules: readonly Module<MixedOutput>[], debugIdModule: Module<VirtualJSOutput>): readonly Module<MixedOutput>[] {
+  const modifiedPreModules = [...preModules];
+  if (modifiedPreModules.length > 0 && modifiedPreModules[0] !== undefined && modifiedPreModules[0].path === PRELUDE_MODULE_PATH) {
+    // prelude module must be first as it measures the bundle startup time
+    modifiedPreModules.unshift(preModules[0]);
+    modifiedPreModules[1] = debugIdModule;
+  } else {
+    modifiedPreModules.unshift(debugIdModule);
+  }
+  return modifiedPreModules;
+}
+
+async function extractSerializerResult(serializerResult: MetroSerializerOutput): Promise<SerializedBundle> {
+  let code: string;
+  let map: string = '{}';
+
+  if (typeof serializerResult === 'string') {
+    code = serializerResult;
+  } else if ('map' in serializerResult) {
+    code = serializerResult.code;
+    map = serializerResult.map;
+  } else {
+    const awaitedResult = await serializerResult;
+    if (typeof awaitedResult === 'string') {
+      code = awaitedResult;
+    } else {
+      code = awaitedResult.code;
+      map = awaitedResult.map;
+    }
+  }
+
+  return { code, map };
```

Comment:
let's avoid mutating code/map in favour of multiple return statements.

```suggestion
  if (typeof serializerResult === 'string') {
    return { code: serializerResult, map: '{}' };
   }
  
  if ('map' in serializerResult) {
    return { code: serializerResult.code, map: serializerResult.map };
  }
  
  const awaitedResult = await serializerResult;
  if (typeof awaitedResult === 'string') {
    return { code: awaitedResult, map: '{}' };
  }
  
  return { code: awaitedResult.code, map: awaitedResult.map };
```